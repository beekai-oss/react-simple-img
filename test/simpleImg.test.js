import React from 'react';
import renderer from 'react-test-renderer';
import { shallow, mount } from 'enzyme';
import SimpleImg from '../src/simpleImg';

const props = {
  src: 'src',
  className: 'className',
  alt: 'alt',
  sizes: 'sizes',
  srcSet: 'srcSet',
  animationDuration: 1,
  animationEndStyle: { opacity: 0 },
};
const addEventListener = window.addEventListener;
const removeEventListener = window.removeEventListener;

describe('SimpleImg', () => {
  beforeEach(() => {
    jest.resetAllMocks();
    window.addEventListener = jest.fn();
    window.removeEventListener = jest.fn();
    window.__REACT_SIMPLE_IMG__ = {
      observer: {
        observe: jest.fn(),
      },
    };
  });

  afterEach(() => {
    window.addEventListener = addEventListener;
    window.removeEventListener = removeEventListener;
    window.__REACT_SIMPLE_IMG__ = undefined;
  });

  it('should render correctly', () => {
    const tree = renderer.create(<SimpleImg {...props} />);
    expect(tree).toMatchSnapshot();
  });

  it('should render out cached img', () => {
    const tree = shallow(<SimpleImg {...props} />);
    tree.setState({
      isCached: true,
    });
    expect(tree).toMatchSnapshot();
  });

  it('should apply aspect ratio when height and width is supplied', () => {
    const tree = renderer.create(<SimpleImg {...{ ...props, applyAspectRatio: true }} />);
    expect(tree).toMatchSnapshot();
  });

  it('should render background color', () => {
    const tree = renderer.create(<SimpleImg {...{ ...props, placeholder: 'red' }} />);
    expect(tree).toMatchSnapshot();
  });

  it('should render out placeholder as image', () => {
    const tree = renderer.create(<SimpleImg {...{ ...props, placeholder: '/test/image.jpg' }} />);
    expect(tree).toMatchSnapshot();
  });

  it('should render out as image data', () => {
    const tree = renderer.create(<SimpleImg {...{ ...props, placeholder: 'data:image//test' }} />);
    expect(tree).toMatchSnapshot();
  });

  it('should render only span when place holder src is not supplied', () => {
    const { placeholder, ...omitProps } = props;
    const tree = renderer.create(<SimpleImg {...omitProps} />);
    expect(tree).toMatchSnapshot();
  });

  it('should remove window object reference when component removed', () => {
    const unobserveSpy = jest.fn();
    const hasSpy = jest.fn();
    window.__REACT_SIMPLE_IMG__ = {
      observer: {
        observe: () => {},
        unobserve: unobserveSpy,
      },
      imgLoadingRefs: {
        set: () => {},
        delete: () => {},
        has: hasSpy,
      },
    };
    const tree = shallow(<SimpleImg {...{ ...props, useContext: false }} />);
    tree.setState({
      useContext: false,
    });
    tree.instance().element = {
      current: 'test',
    };
    tree.unmount();
    expect(unobserveSpy).toHaveBeenCalled();
    expect(hasSpy.mock.calls[0][0]).toEqual('test');

    window.__REACT_SIMPLE_IMG__ = undefined;
  });

  it('should update isDocumentLoad state', () => {
    const tree = shallow(<SimpleImg {...props} />);
    tree.instance().setDocumentLoaded();
    expect(tree.state('isDocumentLoad')).toBeTruthy();
  });

  it('should remove image refs on componentWillUnmount', () => {
    const deleteSpy = jest.fn();
    const getSpy = jest.fn();

    window.__REACT_SIMPLE_IMG__ = {
      observer: {
        observe: () => {},
        unobserve: () => {},
      },
      imgLoadingRefs: {
        set: () => {},
        delete: deleteSpy,
        get: getSpy,
        has: () => true,
      },
    };

    const tree = shallow(<SimpleImg {...{ ...props, useContext: false }} />);
    const instance = tree.instance();
    getSpy.mockReturnValueOnce({
      src: 'test',
    });
    instance.element = {
      current: 1,
    };
    instance.componentWillUnmount();

    expect(getSpy).toBeCalled();
    expect(deleteSpy).toBeCalled();
  });

  it('should set state to be is cached when image ref is stored in sessionStorage', () => {
    window.sessionStorage.setItem('__REACT_SIMPLE_IMG__', JSON.stringify({ src: 'test' }));
    window.__REACT_SIMPLE_IMG__ = {
      disableAnimateCachedImg: true,
    };
    const tree = mount(<SimpleImg {...props} />);
    expect(tree.state('isCached')).toBeTruthy();
  });
});
