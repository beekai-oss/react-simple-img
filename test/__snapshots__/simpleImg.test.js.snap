// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SimpleImg should apply aspect ratio when height and width is supplied 1`] = `
<div
  className="className"
  style={
    Object {
      "alignItems": "center",
      "display": "flex",
      "flexDirection": "column",
      "height": 1,
      "justifyContent": "center",
      "overflow": "hidden",
      "position": "relative",
      "visibility": "hidden",
    }
  }
>
  <img
    alt="alt"
    data-end-style="opacity:0;"
    data-src="src"
    data-srcset="srcSet"
    sizes="sizes"
    src="data:image/gif;base64,R0lGODlhAQABAAD/ACwAAAAAAQABAAACADs="
    srcSet=""
    style={
      Object {
        "visibility": "hidden",
        "width": "100%",
      }
    }
  />
  <div
    style={
      Object {
        "background": "white",
        "height": "100%",
        "left": 0,
        "position": "absolute",
        "top": 0,
        "transition": "1s all",
        "width": "100%",
      }
    }
  />
</div>
`;

exports[`SimpleImg should render background color 1`] = `
<div
  className="className"
  style={
    Object {
      "alignItems": "center",
      "display": "flex",
      "flexDirection": "column",
      "height": 1,
      "justifyContent": "center",
      "overflow": "hidden",
      "position": "relative",
      "visibility": "hidden",
    }
  }
>
  <img
    alt="alt"
    data-end-style="opacity:0;"
    data-src="src"
    data-srcset="srcSet"
    sizes="sizes"
    src="data:image/gif;base64,R0lGODlhAQABAAD/ACwAAAAAAQABAAACADs="
    srcSet=""
    style={
      Object {
        "visibility": "hidden",
        "width": "100%",
      }
    }
  />
  <div
    style={
      Object {
        "background": "red",
        "height": "100%",
        "left": 0,
        "position": "absolute",
        "top": 0,
        "transition": "1s all",
        "width": "100%",
      }
    }
  />
</div>
`;

exports[`SimpleImg should render correctly 1`] = `
<div
  className="className"
  style={
    Object {
      "alignItems": "center",
      "display": "flex",
      "flexDirection": "column",
      "height": 1,
      "justifyContent": "center",
      "overflow": "hidden",
      "position": "relative",
      "visibility": "hidden",
    }
  }
>
  <img
    alt="alt"
    data-end-style="opacity:0;"
    data-src="src"
    data-srcset="srcSet"
    sizes="sizes"
    src="data:image/gif;base64,R0lGODlhAQABAAD/ACwAAAAAAQABAAACADs="
    srcSet=""
    style={
      Object {
        "visibility": "hidden",
        "width": "100%",
      }
    }
  />
  <div
    style={
      Object {
        "background": "white",
        "height": "100%",
        "left": 0,
        "position": "absolute",
        "top": 0,
        "transition": "1s all",
        "width": "100%",
      }
    }
  />
</div>
`;

exports[`SimpleImg should render only span when place holder src is not supplied 1`] = `
<div
  className="className"
  style={
    Object {
      "alignItems": "center",
      "display": "flex",
      "flexDirection": "column",
      "height": 1,
      "justifyContent": "center",
      "overflow": "hidden",
      "position": "relative",
      "visibility": "hidden",
    }
  }
>
  <img
    alt="alt"
    data-end-style="opacity:0;"
    data-src="src"
    data-srcset="srcSet"
    sizes="sizes"
    src="data:image/gif;base64,R0lGODlhAQABAAD/ACwAAAAAAQABAAACADs="
    srcSet=""
    style={
      Object {
        "visibility": "hidden",
        "width": "100%",
      }
    }
  />
  <div
    style={
      Object {
        "background": "white",
        "height": "100%",
        "left": 0,
        "position": "absolute",
        "top": 0,
        "transition": "1s all",
        "width": "100%",
      }
    }
  />
</div>
`;

exports[`SimpleImg should render out as image data 1`] = `
<div
  className="className"
  style={
    Object {
      "alignItems": "center",
      "display": "flex",
      "flexDirection": "column",
      "height": 1,
      "justifyContent": "center",
      "overflow": "hidden",
      "position": "relative",
      "visibility": "hidden",
    }
  }
>
  <img
    alt="alt"
    data-end-style="opacity:0;"
    data-src="src"
    data-srcset="srcSet"
    sizes="sizes"
    src="data:image//test"
    srcSet=""
    style={
      Object {
        "width": "100%",
      }
    }
  />
  <img
    sizes="sizes"
    src="data:image//test"
    style={
      Object {
        "height": "100%",
        "left": 0,
        "position": "absolute",
        "top": 0,
        "transition": "1s all",
        "width": "100%",
      }
    }
  />
</div>
`;

exports[`SimpleImg should render out cached img 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <SimpleImg
    alt="alt"
    animationDuration={1}
    animationEndStyle={
      Object {
        "opacity": 0,
      }
    }
    className="className"
    importance="low"
    placeholder="white"
    sizes="sizes"
    src="src"
    srcSet="srcSet"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <img
        alt="alt"
        sizes="sizes"
        src="src"
        srcSet="srcSet"
        style={
          Object {
            "width": "100%",
          }
        }
      />,
      "className": "className",
      "style": Object {
        "alignItems": "center",
        "display": "flex",
        "flexDirection": "column",
        "justifyContent": "center",
        "overflow": "hidden",
        "position": "relative",
      },
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "alt": "alt",
        "sizes": "sizes",
        "src": "src",
        "srcSet": "srcSet",
        "style": Object {
          "width": "100%",
        },
      },
      "ref": null,
      "rendered": null,
      "type": "img",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <img
          alt="alt"
          sizes="sizes"
          src="src"
          srcSet="srcSet"
          style={
            Object {
              "width": "100%",
            }
          }
        />,
        "className": "className",
        "style": Object {
          "alignItems": "center",
          "display": "flex",
          "flexDirection": "column",
          "justifyContent": "center",
          "overflow": "hidden",
          "position": "relative",
        },
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "alt": "alt",
          "sizes": "sizes",
          "src": "src",
          "srcSet": "srcSet",
          "style": Object {
            "width": "100%",
          },
        },
        "ref": null,
        "rendered": null,
        "type": "img",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;

exports[`SimpleImg should render out placeholder as image 1`] = `
<div
  className="className"
  style={
    Object {
      "alignItems": "center",
      "display": "flex",
      "flexDirection": "column",
      "height": 1,
      "justifyContent": "center",
      "overflow": "hidden",
      "position": "relative",
      "visibility": "hidden",
    }
  }
>
  <img
    alt="alt"
    data-end-style="opacity:0;"
    data-src="src"
    data-srcset="srcSet"
    sizes="sizes"
    src="/test/image.jpg"
    srcSet=""
    style={
      Object {
        "width": "100%",
      }
    }
  />
  <img
    sizes="sizes"
    src="/test/image.jpg"
    style={
      Object {
        "height": "100%",
        "left": 0,
        "position": "absolute",
        "top": 0,
        "transition": "1s all",
        "width": "100%",
      }
    }
  />
</div>
`;
